### DISCUSSION TITLE
Find what exceptions are raised

### DISCUSSION THREAD

**Discussion by brianreinhold** [pyusb/pyusb#431] [Q&A] [answered]
I am calling bulk endpoint read and write functions, however I am finding it difficult to determine what exceptions these functions raise from the documentation. It's probably somewhere; I keep getting referred to Device.write or Device.Read methods which I suppose means the read and write function in the Device class. However, it's not there either.

Where would I find it?

**Comment by jonasmalacofilho** [comment, marked as answer]
The specific exceptions a method may raise are only documented in a few critical cases. 

In all other ones you'll see a `usb.core.USBError` (or one of its subclasses) for USB-specific errors; and, when it makes sense, you might also get standard Python exceptions (e.g. `ValueError`).

Also note that, generally, what you're interested in is the underlying error reported by the backend (and wrapped into the `USBError`), which is outside our control and can also vary by backend _and_ platform.  You'll need to look into those if exhaustively trying to handle all possible error cases (and, even then, LibUSB 1.0 is allowed to return `LIBUSB_ERROR_OTHER` most [if not all] of the time)...

**Reply by brianreinhold** [reply]
I think I am most interested in the exception that I will get when the connection is lost, such as someone pulling the cable while the device was being read.
